2996_reflection/reflection.md

```diff
@@ -1900,18 +1912,24 @@ namespace std::meta {
 ```
 :::
 
-If `r` is a reflection for a constant-expression or a constant-valued entity of type `T`, `value_of<T>(r)` evaluates to that constant value.
+If `r` is a reflection of a constant-expression or a constant-valued entity of type `T`, `value_of<T>(r)` evaluates to that constant value.
 
-If `r` is a reflection for a variable of non-reference type `T`, `value_of<T&>(r)` and `value_of<T const&>(r)` are lvalues referring to that variable.
-If the variable is usable in constant expressions [expr.const], `value_of<T>(r)` evaluates to its value.
+If `r` is a reflection of a variable of non-reference type `T`, then `value_of<T&>(r)` and `value_of<T const&>(r)` are lvalues referring to that variable.
+If the variable is usable in constant expressions [expr.const], then `value_of<T>(r)` evaluates to its value.
 
-If `r` is a reflection for a variable of reference type `T` usable in constant-expressions, `value_of<T>(r)` evaluates to that reference.
+If `r` is a reflection of a variable of reference type `T` usable in constant-expressions, then `value_of<T>(r)` evaluates to that reference.
 
-If `r` is a reflection of an enumerator constant of type `E`, `value_of<E>(r)` evaluates to the value of that enumerator.
+If `r` is a reflection of a function, or pointer to a function, of type `R(A_0, ... A_n)`, then `value_of<R(*)(A_0, ..., A_n)>(r)` evaluates to a pointer to that function.
 
-If `r` is a reflection of a non-bit-field non-reference non-static member of type `M` in a class `C`, `value_of<M C::*>(r)` is the pointer-to-member value for that nonstatic member.
+If `r` is a reflection of a non-static member function, or pointer to a non-static member function, and `T` is the type for a pointer to the reflected member function, then `value_of<T>(r)` evaluates to a pointer to the member function.
 
-For other reflection values `r`, `value_of<T>(r)` is ill-formed.
+If `r` is a reflection of a lambda expression, or the closure object of a lambda expression, and the entity reflected by `r` is convertible to `T`, then `value_of<T>(r)` evaluates to `static_cast<T>(r)`.
```

Comment:
Er, "convertible"? All of the other rules here are exact. So like `value_of<long>(^0)` is ill-formed, so I think `value_of<void(*)()>(^[]{})` should also be? Or at least, these should be equivalent I think.

Reply:
Hrm. That would require `value_of<decltype(fn)>(fn)`, which is no good if you're getting a reflection of the lambda as an argument :-\

I don't think it's so bad to have a special case for lambdas. I spelled it as-is to try to allow for generic lambas.

If there's tighter wording we could apply to "convertible", I'm open to suggestions. Also open to revisiting what conversions we allow for other types.

Reply:
Don't we still need the sentence that is deleted though?

Reply:
Didn't mean to delete the "Otherwise" clause; thanks.

Re: lambdas, as per offline discussion, pandas are in, generic pandas are out. :panda_face:

Reply:
Wait what? Why not generic lambdas? I _definitely_ dislike _just_ non-generic lambdas.

Reply:
@daveedvdv can confirm, but my understanding is that generic lambdas require emitting AST nodes, where as non-generic lambdas, we can probably pull out the underlying function without needing to. That was the whole "pandas can eat bamboo" remark lol

Reply:
I think my strong preference here is simply no conversions allowed (except for qualification conversions, which nobody actually thinks of as conversions anyway).

If we're going to allow _any_ conversions at all, then I would like that decision to be based upon user-facing C++ rules rather than compiler implementation details. Like, as a user, I don't care about AST nodes - I care about understanding the language rules. So like, it would make sense for me if:

* no conversions were allowed, or
* all conversions were allowed, or
* we specifically special-case lambdas as a whole on the premise that otherwise `value_of<T>(lambda_refl)` is basically impossible. But if we special-case lambdas, we should special-case _all_ lambdas not just some lambdas - since the language rule for that conversion is the same for generic and non-generic lambdas, so it should be the same for `value_of` too. 

That's my order of preference right now anyway. I don't like the lambda carve out simply because of the complexity on the user, but I _especially_ dislike the lambda carve out if it doesn't even carve out all lambdas. 

If the starting point here is basically "wait, `value_of` is useless for lambdas unless we do this?" I think it's better to just start with "yes. yes it is." Especially since right now expression reflection is fairly limited.

Reply:
I've removed the lambda case from `value_of` for the time being. We can circle back later.